File: packages/Menu/scan.m
#import "scan.h"

void scanApps(NSMutableArray<NSString *> **outItems,
              NSMutableDictionary<NSString *, NSString *> **outPaths) {
    NSArray *dirs = @[
        @"/System/Applications",
        @"/System/Applications/Utilities",
        @"/System/Library/CoreServices",
        @"/Applications",
        @"/Applications/Utilities",
        [NSHomeDirectory() stringByAppendingPathComponent:@"Applications/Nix"]
    ];
    NSMutableArray *items = [NSMutableArray new];
    NSMutableDictionary *paths = [NSMutableDictionary new];
    NSFileManager *fm = [NSFileManager defaultManager];

    for (NSString *dir in dirs) {
        BOOL isDir;
        if (![fm fileExistsAtPath:dir isDirectory:&isDir] || !isDir) continue;
        NSArray *contents = [fm contentsOfDirectoryAtPath:dir error:nil];
        for (NSString *entry in contents) {
            if (![entry hasSuffix:@".app"]) continue;
            NSString *name = [entry stringByDeletingPathExtension];
            NSString *path = [dir stringByAppendingPathComponent:entry];
            // Skip duplicates (first found wins)
            if (!paths[name]) {
                paths[name] = path;
                [items addObject:name];
            }
        }
    }
    [items sortUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
    *outItems = items;
    *outPaths = paths;
}

File: packages/Menu/Scan.swift
import Foundation

private let appDirs = [
    "/System/Applications",
    "/System/Applications/Utilities",
    "/System/Library/CoreServices",
    "/Applications",
    "/Applications/Utilities",
]

func scanApps() -> (items: [String], paths: [String: String]) {
    var items: [String] = []
    var paths: [String: String] = [:]
    let dirs = appDirs + [NSHomeDirectory() + "/Applications/Nix"]
    let fm = FileManager.default

    for dir in dirs {
        guard let contents = try? fm.contentsOfDirectory(atPath: dir) else { continue }
        for entry in contents where entry.hasSuffix(".app") {
            let name = (entry as NSString).deletingPathExtension
            if paths[name] == nil {
                paths[name] = "\(dir)/\(entry)"
                items.append(name)
            }
        }
    }
    return (items.sorted { $0.localizedCaseInsensitiveCompare($1) == .orderedAscending }, paths)
}
